server/src/main/java/org/elasticsearch/action/bulk/BulkRequestBuilder.java

```diff
@@ -126,32 +159,107 @@ public BulkRequestBuilder setWaitForActiveShards(final int waitForActiveShards)
      * A timeout to wait if the index operation can't be performed immediately. Defaults to {@code 1m}.
      */
     public final BulkRequestBuilder setTimeout(TimeValue timeout) {
-        request.timeout(timeout);
+        this.timeout = timeout;
         return this;
     }
 
     /**
      * A timeout to wait if the index operation can't be performed immediately. Defaults to {@code 1m}.
      */
     public final BulkRequestBuilder setTimeout(String timeout) {
-        request.timeout(timeout);
+        this.timeoutString = timeout;
         return this;
     }
 
     /**
      * The number of actions currently in the bulk.
      */
     public int numberOfActions() {
-        return request.numberOfActions();
+        return requests.size() + requestBuilders.size() + framedDataList.size();
     }
 
     public BulkRequestBuilder pipeline(String globalPipeline) {
-        request.pipeline(globalPipeline);
+        this.globalPipeline = globalPipeline;
         return this;
     }
 
     public BulkRequestBuilder routing(String globalRouting) {
-        request.routing(globalRouting);
+        this.globalRouting = globalRouting;
         return this;
     }
+
+    @Override
+    public BulkRequestBuilder setRefreshPolicy(WriteRequest.RefreshPolicy refreshPolicy) {
+        this.refreshPolicy = refreshPolicy;
+        return this;
+    }
+
+    @Override
+    public BulkRequestBuilder setRefreshPolicy(String refreshPolicy) {
+        this.refreshPolicyString = refreshPolicy;
+        return this;
+    }
+
+    @Override
+    public BulkRequest request() {
+        validate();
+        BulkRequest request = new BulkRequest(globalIndex);
+        if (requests.isEmpty() == false && requestBuilders.isEmpty() == false) {
```

Comment:
Given the similar validation in `validate()` I think this separate check here should be dropped.